File: solutions/lua/roman-numerals/1/roman-numerals.lua
return {
	to_roman = function(number)
		local letters = {{'M', 1000}, {'C', 100}, {'X', 10}, {'I', 1}}
		local cinq = {{}, {'D', 500}, {'L', 50}, {'V', 5}}
		local res = ""
		while number > 0 do
			for i, p in pairs(letters) do
				local l = p[1]
				local v = p[2]
				local m = number // p[2]
				if m == 9 then
					res = res .. l .. letters[i-1][1]
				elseif 5 <= m and m <= 8 then
					res = res .. cinq[i][1] .. string.rep(l, m-5)
				elseif m == 4 then
					res = res .. l .. cinq[i][1]
				elseif 1 <= m and m <= 3 then
					res = res .. string.rep(l, m)
				end
				if m > 0 then
					number = number - m*v
					break
				end
			end
		end
		return res
	end
}

File: solutions/c/roman-numerals/1/roman_numerals.c
#include "roman_numerals.h"

#include <stdlib.h>

typedef struct { char first; int second; } pair;

// using M = std::vector<std::pair<char, int>>;
typedef pair m[];

char *to_roman_numeral(unsigned int number) {
	pair letters[] = {{'M', 1000}, {'C', 100}, {'X', 10}, {'I', 1}};
	pair cinq[] = {{'K', 0}, {'D', 500}, {'L', 50}, {'V', 5}};
	char *res = malloc(sizeof(char) * 500);
	int sz = 0;
	while (number > 0) {
		for (int i = 0ul; i < 4; i++) {
			pair p = letters[i];
			char l = p.first;
			int v = p.second;
			int m = number / v;
			if (m == 9) {
				res[sz++] = l;
				res[sz++] = letters[i-1].first;
			} else if (5 <= m && m <= 8) {
				res[sz++] = cinq[i].first;
				for (int j = 0; j < (m-5); j++)
					res[sz++] = l;
			} else if (m == 4) {
				res[sz++] = l;
				res[sz++] = cinq[i].first;
			} else if (1 <= m && m <= 3) {
				for (int j = 0; j < m; j++)
					res[sz++] = l;
			}
			if (m > 0) {
				number -= m*v;
				break;
			}
		}
	}
	res[sz] = '\0';
	return res;
}
